design/baremetal-operator/kubebuilder-migration.md

```diff
@@ -0,0 +1,213 @@
+4. Use kubebuilder to create an API and controller for BareMetalHost,
+   then replace the implementations with the existing ones.
+5. Restore the rest of the existing code from the pkg directory,
+   moving it to the root of the repository.
+6. Fix everything so it compiles and the tests run.
+7. Add missing targets to the Makefile (lint, sec, openapi generation,
+   etc.).
+
+### Implementation Details/Notes/Constraints
+
+Using `git mv` as often as possible will help link the new files to
+their history in any old locations.
+
+Using an empty string for the domain argument to `kubebuilder init`
+means that we can specify a group of `"metal3.io"` when creating the
+host API, and avoid ending up renaming the resource type. Future APIs
+should include a more well-scoped group ("networking.metal3.io" or
+"hardware.metal3.io") but changing the group name for the existing
+resource will make it exceptionally difficult to upgrade.
+
+### Risks and Mitigations
+
+Although operator-sdk and kubebuilder both use the same underlying
+code generation library, and are trying to achieve parity, they are
+likely to be using different versions and options from time to
+time. The new implementation will therefore introduce updated
+dependencies, and the longer we put off the work the more likely those
+dependencies are to require code changes. We will mitigate some of
+that risk by vendoring a copy of the tool, which lets us lock in a
+specific version to use.
+
+Bug fixes will be more challenging to backport after the migration,
+because the code layout will be different. This risk is mitigated by
+the fact that we do not maintain old versions of the
+baremetal-operator today, and maintain backwards compatibility in all
+changes. The new implementation should also be API-compatible with the
+older implementation. We can create a maintenance branch for anyone
+using the baremetal-operator who wishes to support the older
+implementation for any reason.
+
+Effectively re-implementing something that works is inherently risky.
+Most of the differences in the implementation will be in the generated
+code that invokes the same controller runtime library, though, which
+will minimize the changes in behavior.
+
+Other projects that import the BMO code (CAPM3, CAPBM, etc.) will need
+to update the import statements after upgrading to the new
+implementation. Until the API of the BareMetalHost is changed, these
+updates aren't strictly necessary, but if we do them sooner that will
+make future API changes easier to follow.
+
+### Work Items
+
+1. Tag a release and create a maintenance branch for the
+   baremetal-operator.
+2. Follow a process as described in the Design Details section above
+   to migrate to kubebuilder.
+3. Tag a release to use for including in the cluster API providers.
+4. Rebase all open PRs after the new implementation is approved.
+5. Update the imports in CAPM3.
+6. Update the imports and vendoring in CAPBM.
```

Comment:
Is this really needed considering we deprecated the repo ?

Reply:
I need to do that step downstream where the repo is still used. I can clarify that in the text or remove the task entirely.